src/main/java/com/scalar/database/storage/cassandra4driver/ResultImpl.java

```diff
@@ -0,0 +1,159 @@
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.Immutable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Immutable
+public class ResultImpl implements Result {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResultImpl.class);
+  private final TableMetadata metadata;
+  private Map<String, Value> values;
+
+  public ResultImpl(Row row, TableMetadata metadata) {
+    checkNotNull(row);
+    this.metadata = checkNotNull(metadata);
+    interpret(row);
+  }
+
+  /**
+   * constructor only for testing purposes
+   *
+   * @param values
+   */
+  @VisibleForTesting
+  ResultImpl(Collection<Value> values, TableMetadata metadata) {
+    this.metadata = metadata;
+    this.values = new HashMap<>();
+    values.forEach(v -> this.values.put(v.getName(), v));
+  }
+
+  @Override
+  public Optional<Key> getPartitionKey() {
+    return getKey(metadata.getPartitionKeyNames());
+  }
+
+  @Override
+  public Optional<Key> getClusteringKey() {
+    return getKey(metadata.getClusteringColumnNames());
+  }
+
+  @Override
+  public Optional<Value> getValue(String name) {
+    return Optional.ofNullable(values.get(name));
+  }
+
+  @Override
+  @Nonnull
+  public Map<String, Value> getValues() {
+    return Collections.unmodifiableMap(values);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(values);
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (o == this) {
+      return true;
+    }
+    if (!(o instanceof ResultImpl)) {
+      return false;
+    }
+    ResultImpl other = (ResultImpl) o;
+    if (this.values.equals(other.values)) {
+      return true;
+    }
+    return false;
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this).add("values", values).toString();
+  }
+
+  @VisibleForTesting
+  void interpret(Row row) {
+    values = new HashMap<>();
+    getColumnDefinitions(row)
+        .forEach(
+            (name, type) -> {
+              Value value = convert(row, name, type);
+              values.put(name, value);
+            });
+  }
+
+  @VisibleForTesting
+  Map<String, DataType> getColumnDefinitions(Row row) {
+    Map<String, DataType> definitions = new HashMap<>();
+    row.getColumnDefinitions().forEach(c -> definitions.put(c.getName().toString(), c.getType()));
+    return definitions;
+  }
+
+  private Optional<Key> getKey(Set<String> names) {
+    List<Value> list = new ArrayList<>();
+    for (String name : names) {
+      Value value = values.get(name);
+      if (value == null) {
+        LOGGER.warn("full key doesn't seem to be projected into the result");
+        return Optional.empty();
+      }
+      list.add(value);
+    }
+    return Optional.of(new Key(list));
+  }
+
+  private static Value convert(Row row, String name, DataType type)
+      throws UnsupportedTypeException {
+    int typeCode = type.getProtocolCode();
+    if (typeCode == DataTypes.BOOLEAN.getProtocolCode()) {
```

Comment:
Just a thought, but would this look nicer as a `switch` statement?

Reply:
Thanks!
I thought about that. But I tried it and got `constant expression required` error.